Add build number to Arm compiler version identification

LPettey-Arm opened the pull request:
Arm, Inc. intends to use Spack as part of new internal CI process for HPC applications. As such, we would like Spack to identify the build number of the compiler as well as the major, minor and minor minor version numbers. This change will not affect any other compiler version identification. Thanks!

adamjstewart commented:
This change looks good to me! Thanks for choosing Spack!

Flake8 changes needed:
```
lib/spack/spack/compilers/arm.py:38: [E501] line too long (95 > 79 characters)
lib/spack/spack/compilers/arm.py:39: [W293] blank line contains whitespace
lib/spack/spack/compilers/arm.py:41: [W293] blank line contains whitespace
lib/spack/spack/compilers/arm.py:42: [E304] blank lines found after function decorator
lib/spack/spack/compilers/arm.py:44: [W293] blank line contains whitespace
lib/spack/spack/compilers/arm.py:45: [F821] undefined name 're'
lib/spack/spack/compilers/arm.py:54: [E301] expected 1 blank line, found 0
```
Also, can you update the tests in the `test_arm_version_detection` function of  `lib/spack/spack/test/compilers.py`? Let me know if you need help.

becker33 commented on lib/spack/spack/compilers/arm.py:
```diff
@@ -35,8 +35,22 @@ class Arm(spack.compiler.Compiler):
+    def extract_version_from_output(cls, output):
+        """Extracts the version from compiler's output."""
+        
+        match = re.search(cls.version_regex, output)
```
`re` hasn't been imported.

LPettey-Arm commented:
Thanks for looking at this so quickly! I think I've cleaned up the fleck8 errors but I'll hold on the commit until I look into the `test_arm_version_detection` function. Here is the version reporting information.

Before suggested patch:
```
[ec2-user@ip-172-31-7-135 spack]$ armclang --version
Arm C/C++/Fortran Compiler version 20.0 (build number 51) (based on LLVM 9.0.1)
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/ec2-user/armcc/arm-linux-compiler-20.0_Generic-AArch64_RHEL-7_aarch64-linux/bin

[ec2-user@ip-172-31-7-135 spack]$ spack compiler find
==> Added 1 new compiler to /home/ec2-user/.spack/linux/compilers.yaml
    arm@20.0
==> Compilers are defined in the following files:
    /home/ec2-user/.spack/linux/compilers.yaml
```
After suggested patch:
```
[ec2-user@ip-172-31-7-135 spack]$ armclang --version
Arm C/C++/Fortran Compiler version 20.0 (build number 51) (based on LLVM 9.0.1)
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/ec2-user/armcc/arm-linux-compiler-20.0_Generic-AArch64_RHEL-7_aarch64-linux/bin

[ec2-user@ip-172-31-7-135 spack]$ spack compiler find
==> Added 1 new compiler to /home/ec2-user/.spack/linux/compilers.yaml
    arm@20.0.0.51
==> Compilers are defined in the following files:
    /home/ec2-user/.spack/linux/compilers.yaml
```

LPettey-Arm commented:
I've updated my fork so the flake8 tests pass.

I also fixed the `test_arm_version_detection` function to support the 4 digit output. 

Please let me know if there's anything else I need to do for this pull request. Thanks!

adamjstewart commented on lib/spack/spack/compilers/arm.py:
```diff
@@ -35,7 +36,21 @@ class Arm(spack.compiler.Compiler):
+        if match:
+            if match.group(1).count('.') == 1:
+                temp = match.group(1) + ".0." + match.group(2)
+            else:
```
Can you add a second test parameterization where this else clause gets hit?

LPettey-Arm replied:
The Arm compiler is generally released twice per year. June for ISC and November for SC. These releases are either a major (e.g. 20.0) or minor (20.1) release. The rare occasion does arise that there is a release between these events and a minor minor version is justified (e.g. 20.1.1).

When discussing the build feature detection with the compiler team, they opted for this approach in case a minor minor version is needed. None of the current releases have a minor minor designation, but this provides flexibility for the future.

adamjstewart replied:
In that case, maybe you can add a fake test so we know it works and get test coverage.

LPettey-Arm replied:
I added a fake test so if the version output is "...19.3.2 (build number 75)..." the test result is "19.3.2.75" consistent with the new version function. Both the 'if' and 'else' now have a sample test.

adamjstewart commented on lib/spack/spack/compilers/arm.py:
```diff
@@ -35,7 +36,21 @@ class Arm(spack.compiler.Compiler):
+                temp = match.group(1) + ".0." + match.group(2)
+            else:
+                temp = match.group(1) + "." + match.group(2)
+        else:
```
Is there actually a situation where this else clause will be hit?

alalazo replied:
The `else` clause can be avoided completely if we write things like:
```python
match = re.search(cls.version_regex, output)
temp = 'unknown'
if match:
    ...

return temp
```

LPettey-Arm replied:
I can't think of a case where the compiler is found but there is not version number. 

I can re-write the compiler version function to eliminate the else clause as @alalazo suggests, but its functionality will be the same. None of the compiler version tests are written with a situation where "unknown" is returned. 

Please advise if you would like changes to the code. Thanks!

adamjstewart replied:
Yes, please make the change. I think we can get your changes to 100% test coverage, which will help me sleep at night 😄

LPettey-Arm replied:
I changed the compiler version function to @alalazo 's suggestion. There is no 'else' clause for the 'if match' check. Thanks for all the feedback!
